### PR TITLE
[6.3] fix enhance ssh timeout functionality

### PR DESCRIPTION
**Pull request by ldjebran**
The existing timeout was confused with connection timeout, that was also passed as second argument to wrong client function arg. this caused some hammer hanging commands to block the automation tests as also had no timeout.

example of the implemented timeout when the test launch a blocking cmd
fromthe test log:
```console
2017-07-19 11:14:07 - robottelo.ssh - INFO - >>> LANG=en_US.UTF-8  hammer -v -u admin -p changeme --output=csv host package-group remove --groups="birds" --host-id="52"
2017-07-19 11:16:07 - robottelo.ssh - ERROR - ssh command timeout 120 reached
```
the produced error:
```console
tests/foreman/cli/test_host.py:1913: in test_positive_remove_package
    u'packages': FAKE_0_CUSTOM_PACKAGE_NAME,
robottelo/cli/host.py:110: in package_remove
    cls._construct_command(options), output_format='csv')
robottelo/cli/base.py:299: in execute
    conn_timeout=conn_timeout,
robottelo/ssh.py:270: in command
    cmd, connection, output_format, timeout, conn_timeout)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cmd = 'LANG=en_US.UTF-8  hammer -v -u admin -p changeme --output=csv host package remove --packages="bear" --host-id="51"'
connection = <robottelo.ssh.SSHClient object at 0x7f1b9ca57d50>, output_format = 'csv', timeout = 120, conn_timeout = 10

    def execute_command(cmd, connection, output_format=None, timeout=120,
                        conn_timeout=10):
        """Execute a command via ssh in the given connection
    
        :param cmd: a command to be executed via ssh
        :param connection: SSH Paramiko client connection
        :param output_format: plain|json|csv|list valid only for hammer commands
        :param timeout: Time to wait for the ssh command to finish.
        :param conn_timeout: Time to wait for establishing the connection.
        :return: SSHCommandResult
        """
        logger.info('>>> %s', cmd)
        _, stdout, stderr = connection.exec_command(cmd, timeout=conn_timeout)
        if timeout:
            # wait for the exit status ready
            end_time = time.time() + timeout
            while time.time() < end_time:
                if stdout.channel.exit_status_ready():
                    break
                time.sleep(1)
            else:
                logger.error('ssh command timeout %s reached', timeout)
                raise SSHCommandTimeoutError(
>                   'ssh cmd timeout {0} reached'.format(timeout))
E               SSHCommandTimeoutError: ssh cmd timeout 120 reached

robottelo/ssh.py:296: SSHCommandTimeoutError
```
please let know if you need the full test log
Thanks to @abalakh that proposed the solution

### CODE REVIEW

File: robottelo/cli/base.py [hunk 1]
```diff
@@ -275,7 +275,8 @@ def _get_username_password(cls, username=None, password=None):
 
     @classmethod
     def execute(cls, command, user=None, password=None, output_format=None,
-                timeout=None, ignore_stderr=None, return_raw_response=None):
+                timeout=120, ignore_stderr=None, return_raw_response=None,
+                conn_timeout=10):
```
Comment (abalakh) [inline robottelo/cli/base.py]: As these two timeouts are already confusing, i'd suggest to name this one `connection_timeout` to make it absolutely clear.

File: robottelo/cli/contentview.py [hunk 2]
```diff
@@ -117,7 +117,7 @@ def publish(cls, options, timeout=None):
         cls.command_sub = 'publish'
         # Publishing can take a while so try to wait a bit longer
         if timeout is None:
-            timeout = 120
+            timeout = 600
```
Comment (abalakh) [inline robottelo/cli/contentview.py]: Why 600? For api we have 300 secs for everything and 900 for few commands, affected by some BZ. I think we should use the same approach here not to introduce differences in api/cli behavior. And btw, `publish` is not the only command which takes more than 2 minutes, please update others accordingly.

File: robottelo/ssh.py [hunk 3]
```diff
@@ -13,6 +15,10 @@
 logger = logging.getLogger(__name__)
 
 
+class SSHCommandTimeoutError(Exception):
+    """Raised when the SSH command timeout."""
```
Comment (abalakh) [inline robottelo/ssh.py]: I'd rephrase it, as there're few possible timeouts. `...when executed SSH command is not returning any results for certain period of time` maybe?

File: robottelo/ssh.py [hunk 4]
```diff
@@ -237,7 +243,7 @@ def download_file(remote_file, local_file=None, hostname=None):
 
 
 def command(cmd, hostname=None, output_format=None, username=None,
-            password=None, key_filename=None, timeout=10):
+            password=None, key_filename=None, timeout=120, conn_timeout=10):
```
Comment (abalakh) [inline robottelo/ssh.py]: I'm not sure whether 120 secs will be enough. API is returning results immediately, but for some requests which schedule tasks we're waiting for 300 sec to finish by default. The thing is CLI behavior is different and usually we don't get a result immediately but only after corresponding task has finished. 
I can see 2 possible approaches here: 
1) Set default timeout to 300 secs, so max timeout will be the same as in API, but we may start waiting too much for 'atomic' commands to finish (in case they stop responding, which is very unlikely though);
2) Leave 120 secs but set custom 300 secs for all the commands which are scheduling tasks and taking longer to finish, but the tricky part here is to form a list of such commands.

File: robottelo/ssh.py [hunk 5]
```diff
@@ -253,26 +259,41 @@ def command(cmd, hostname=None, output_format=None, username=None,
...
+        # wait for the exit status ready
+        end_time = time.time() + timeout
+        while time.time() < end_time:
+            if stdout.channel.exit_status_ready():
+                break
+            time.sleep(1)
+        else:
+            logger.error('ssh command timeout %s reached', timeout)
+            raise SSHCommandTimeoutError(
+                'ssh cmd timeout {0} reached'.format(timeout))
```
Comment (abalakh) [inline robottelo/ssh.py]: Please rephrase both logger message and exception message. Someone who's analyzing test failures and who's not aware of this PR should be able to understand the fact that test failed because some cli command (btw why not to include specific command in message?) 'hanged' and stopped responding, not just some 'timeout' happened, we have few different timeouts which may occur.

File: robottelo/ssh.py (hunk 4 above)
Comment (ldjebran) [inline robottelo/ssh.py]: this is a preventive functionality from hanging commands, so we can put as much as we can (with a reasonable value of course), so will put 300 everywhere to be compatible with API

File: robottelo/ssh.py (hunk 5 above)
Comment (ldjebran) [inline robottelo/ssh.py]: will put the command in exception, but in logging we have the command already logged, when launched

**Comment by ldjebran** [comment]
@abalakh made the needed changes 
test log: https://url.corp.redhat.com/fix-enhance-ssh-timeout-functionality
the exception at line: 1150
log about timeout at line :  2303

**Comment by ldjebran** [comment]
@abalakh removed one cleanup from test_hosts KatelloAgent TestCase
test-log:  https://url.corp.redhat.com/6-3-fix-enhance-ssh-timeout-functionality-3

**Comment by ldjebran** [comment]
in the test log above notice the exception at line 98 and the corresponding logging error entry at  line 1166

**Comment by sghai** [comment]
ACK.. changes looks good to me..

**Comment by sghai** [comment]
@oshtaier : would you please review the PR ?

File: robottelo/ssh.py [hunk 6]
```diff
@@ -253,26 +262,45 @@ def command(cmd, hostname=None, output_format=None, username=None,
...
-    :param timeout: defaults to 120
+    :param timeout: Time to wait for the ssh command to finish.
+    :param connection_timeout: Time to wait for establishing the connection.
     :return: SSHCommandResult
     """
     logger.info('>>> %s', cmd)
-    _, stdout, stderr = connection.exec_command(cmd, timeout)
+    _, stdout, stderr = connection.exec_command(
+        cmd, timeout=connection_timeout)
+    if timeout:
```
Comment (oshtaier) [inline robottelo/ssh.py]: What are we waiting for here? Timeout is already triggered or not?

**Comment by oshtaier** [comment]
Maybe we can move `timeout` and `conntection_timeout` into properties file with predefined values set to 300 and 10 correspondingly?

File: robottelo/ssh.py (hunk 6 above)
Comment (ldjebran) [inline robottelo/ssh.py]: we are waiting for the exit status of the ssh command

**Comment by ldjebran** [comment]
@oshtaier  sure we can move that values to properties files, but why only for CLI, perhaps it's an issue for a separate issue/PR

**Comment by oshtaier** [comment]
it is not for CLI, but for all SSH connections and as that PR focused on this, so definitely it is appropriate to perform steps in scope of it

**Comment by abalakh** [comment]
@oshtaier but we don't have such timeouts for UI or API in properties file and i'd rather agree with @ldjebran  it's a separate issue to put all timeouts in order.

**Comment by oshtaier** [comment]
it is SSH timeout, not CLI

**Comment by abalakh** [comment]
Right, but my point was control over ssh timeout via properties file is good, same for cli commands timeout, same for api tasks timeout, same for general ui waiter timeout, and it sounds like a separate task for me to move all of them. But that's just my imo
One can even argue with us as i've heard multiple times that our properties file is too big and we need to stop moving all the variables there

**Comment by ldjebran** [comment]
@oshtaier this is a fix PR and not trying to introduce new functionality, we can open an issue for that and do it later, but if you insist we can realize that in this PR.

**Comment by oshtaier** [comment]
You are not fixing, but more improving things. Anyway, you are introducing new constants and I think it is better to store anything related to global timeouts in property file. Code should not be changed at all, but `300` and `10` moved to file

**Comment by ldjebran** [comment]
@oshtaier @sghai @abalakh comments addressed
test log: https://privatebin-it-iso.int.open.paas.redhat.com/?4aec612989048071#DPbBtARUwidIZyXxV/2J3mFGyUJE9RCfPnDFSb72x40=

**Comment by oshtaier** [comment]
Not sure that I understand point at that moment. We moved timeouts to property file, so why do we still have static timeouts in the code?

**Comment by ldjebran** [comment]
yes that one on publish has no sense now that we have by default 300 seconds, will update with unittests

**Comment by ldjebran** [comment]
@oshtaier changes done:
portion of the test failure 
```
test_host.py:1787: in setUp
    KatelloAgentTestCase.activation_key['name'],
../../../robottelo/vm.py:344: in register_contenthost
    result = self.run(cmd)
../../../robottelo/vm.py:391: in run
    return ssh.command(cmd, hostname=self.ip_addr)
../../../robottelo/ssh.py:284: in command
    cmd, connection, output_format, timeout, connection_timeout)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cmd = 'subscription-manager register --org xzDoYO --activationkey ofgkt7rMwX --force'
connection = <robottelo.ssh.SSHClient object at 0x7ff98e5121d0>
output_format = None, timeout = 300, connection_timeout = 10

    def execute_command(cmd, connection, output_format=None, timeout=None,
                        connection_timeout=None):
        """Execute a command via ssh in the given connection
    
        :param cmd: a command to be executed via ssh
        :param connection: SSH Paramiko client connection
        :param output_format: plain|json|csv|list valid only for hammer commands
        :param timeout: Time to wait for the ssh command to finish.
        :param connection_timeout: Time to wait for establishing the connection.
        :return: SSHCommandResult
        """
        if timeout is None:
            timeout = settings.ssh_client.command_timeout
        if connection_timeout is None:
            connection_timeout = settings.ssh_client.connection_timeout
        logger.info('>>> %s', cmd)
        _, stdout, stderr = connection.exec_command(
            cmd, timeout=connection_timeout)
        if timeout:
            # wait for the exit status ready
            end_time = time.time() + timeout
            while time.time() < end_time:
                if stdout.channel.exit_status_ready():
                    break
                time.sleep(1)
            else:
                logger.error('ssh command did not respond in the predefined time'
                             ' (timeout=%s) and will be interrupted', timeout)
                raise SSHCommandTimeoutError(
                    'ssh command: {0} \n did not respond in the predefined time '
>                   '(timeout={1})'.format(cmd, timeout)
                )
E               SSHCommandTimeoutError: ssh command: subscription-manager register --org xzDoYO --activationkey ofgkt7rMwX --force 
E                did not respond in the predefined time (timeout=300)

../../../robottelo/ssh.py:317: SSHCommandTimeoutError
``` 
and log:
```
2017-07-25 17:44:24 - robottelo.ssh - DEBUG - Destroyed Paramiko client 0x7ff98e512d10
2017-07-25 17:44:25 - robottelo.ssh - DEBUG - Instantiated Paramiko client 0x7ff98e5121d0
2017-07-25 17:44:25 - robottelo.ssh - INFO - Connected to [10.8.29.62]
2017-07-25 17:44:25 - robottelo.ssh - INFO - >>> subscription-manager register --org xzDoYO --activationkey ofgkt7rMwX --force
2017-07-25 17:49:26 - robottelo.ssh - ERROR - ssh command did not respond in the predefined time (timeout=300) and will be interrupted
2017-07-25 17:49:26 - robottelo.ssh - DEBUG - Destroyed Paramiko client 0x7ff98e5121d0
2017-07-25 17:49:27 - robottelo.ssh - DEBUG - Instantiated Paramiko client 0x7ff98e51cb90
2017-07-25 17:49:27 - robottelo.ssh - INFO - Connected to [XXXXXXXXXXXXX]
2017-07-25 17:49:27 - robottelo.ssh - INFO - >>> virsh destroy 140709811547152.XXXXXXXXXXXXXXX
2017-07-25 17:49:34 - robottelo.ssh - INFO - <<< stdout
Domain 140709811547152.XXXXXXXXXXXXXXX destroyed
```

**Comment by oshtaier** [comment]
@abalakh please re-review the PR

File: robottelo.properties.sample [hunk 7]
```diff
@@ -25,6 +25,13 @@ ssh_key=
 # Admin password when accessing API and UI
 # admin_password=changeme
 
+# section for ssh client settings
+[ssh_client]
```
Comment (abalakh) [inline robottelo.properties.sample]: And because there's no validation you've missed an issue here :) It should be commented by default - all the custom sections are either fully commented out or fully uncommented

File: robottelo.properties.sample [hunk 8]
```diff
@@ -25,6 +25,13 @@ ssh_key=
...
+# Time to wait for the ssh command to finish, in seconds
+# command_timeout = 300
```
Comment (abalakh) [inline robottelo.properties.sample]: I'd better follow the same style as all the rest properties: no extra spaces, just
```
# command_timeout=300
```

File: robottelo.properties.sample [hunk 9]
```diff
@@ -25,6 +25,13 @@ ssh_key=
...
+# Time to wait for establishing the ssh connection, in seconds
+# connection_timeout = 10
```
Comment (abalakh) [inline robottelo.properties.sample]: Same here

**Comment by abalakh** [comment]
@oshtaier not sure what to tell you :) As for me, separate `[ssh_client]` section is redundant, i'd rather like to see some `[timeouts]` section which covers all out timeouts, not just ssh, as having control over timeouts used for cli but not api or ui seems weird for me. But that's definitely out of scope, and that's why i didn't want extra properties to be implemented as part of this PR.

**Comment by oshtaier** [comment]
not sure how many times I should tell that it is timeouts for ssh, not cli

**Comment by ldjebran** [comment]
@abalakh @oshtaier comments addressed
